src/integration/resources/inprocess/dbnode.go

```diff
@@ -0,0 +1,329 @@
+//
+// The dbnode will start up as you specify in your config. However, there is some
+// helper logic to avoid port and filesystem collisions when spinning up multiple components
+// within the process. If you specify a port of 0 in any address, 0 will be automatically
+// replaced with an open port. This is similar to the behavior net.Listen provides for you.
+// Similarly, for filepaths, if a "*" is specified in the config, then that field will
+// be updated with a temp directory that will be cleaned up when the dbnode is destroyed.
+// This should ensure that many of the same component can be spun up in-process without any
+// issues with collisions.
+func NewDBNode(cfg config.Configuration, opts DBNodeOptions) (resources.Node, error) {
+	// Replace any "0" ports with an open port
+	cfg, err := updateDBNodePorts(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// Replace any "*" filepath with a temporary directory
+	cfg, tmpDirs, err := updateDBNodeFilepaths(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// Configure TChannel client for hitting the DB node.
+	tchanClient, err := integration.NewTChannelClient("client", cfg.DB.ListenAddressOrDefault())
+	if err != nil {
+		return nil, err
+	}
+
+	// Configure logger
+	if opts.Logger == nil {
+		opts.Logger, err = zap.NewDevelopment()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Start the DB node
+	node := &dbNode{
+		cfg:         cfg,
+		logger:      opts.Logger,
+		tchanClient: tchanClient,
+		tmpDirs:     tmpDirs,
+	}
+	node.start()
+
+	return node, nil
+}
+
+func (d *dbNode) start() {
+	interruptCh := make(chan error, d.cfg.Components())
+	shutdownCh := make(chan struct{}, d.cfg.Components())
+	go func() {
+		server.RunComponents(server.Options{
+			Configuration: d.cfg,
+			InterruptCh:   interruptCh,
+			ShutdownCh:    shutdownCh,
+		})
+	}()
+
+	d.interruptCh = interruptCh
+	d.shutdownCh = shutdownCh
+}
+
+func (d *dbNode) HostDetails(port int) (*admin.Host, error) {
+	// TODO(nate): implement once working on helpers for spinning up
+	// a multi-node cluster since that's what it's currently being
+	// used for based on the docker-based implementation.
+	// NOTES:
+	// - id we can generate
+	// - isolation_group set a constant?
+	// - weight just use 1024?
+	// - address is 0.0.0.0
+	// - port is the listen address (get from config)
+
+	return &admin.Host{
+		Id:             "foo",
+		IsolationGroup: "foo-a",
+		Zone:           "embedded",
+		Weight:         1024,
+		Address:        "0.0.0.0",
+		Port:           uint32(port),
+	}, nil
+}
+
+func (d *dbNode) Health() (*rpc.NodeHealthResult_, error) {
+	return d.tchanClient.TChannelClientHealth(defaultRPCTimeout)
+}
+
+func (d *dbNode) WaitForBootstrap() error {
+	return retry(func() error {
+		health, err := d.Health()
+		if err != nil {
+			return err
+		}
+
+		if !health.GetBootstrapped() {
+			err = fmt.Errorf("not bootstrapped")
+			d.logger.Error("node not bootstrapped", zap.Error(err))
```

Comment:
how spammy is this log?

Reply:
Not very. Retry is done with exponential backoff. I've been thinking about an optimization to only dump the stacktrace on the final failure, but don't think we need that just yet.

Reply:
yeah early failure is the best.